### PR TITLE
ceph: Enable external resource script to collect active and standby mgrs

### PR DESCRIPTION
**Pull request by aruniiird**
**Description of your changes:**
Now 'create-external-cluster-resources.py' script will add standby mgrs
along with the active mgr while providing the output.
'MonitoringEndpoint' data-key will contain a comma separated list of
active followed by standby mgrs' IP addresses.

Signed-off-by: Arun Kumar Mohan <amohan@redhat.com>

**Which issue is resolved by this Pull Request:**
Resolves https://github.com/rook/rook/issues/6665

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

File: cluster/examples/kubernetes/ceph/create-external-cluster-resources.py [hunk 1]
```diff
@@ -577,7 +600,7 @@ def _gen_output_map(self):
             self.out_map['CSI_CEPHFS_PROVISIONER_SECRET'] = self.create_cephCSIKeyring_cephFSProvisioner()
         self.out_map['RGW_ENDPOINT'] = self._arg_parser.rgw_endpoint
         self.out_map['MONITORING_ENDPOINT'], \
-            self.out_map['MONITORING_ENDPOINT_PORT'] = self.get_active_ceph_mgr()
+            self.out_map['MONITORING_ENDPOINT_PORT'] = self.get_active_and_standby_mgrs()
```
Comment (leseb) [inline cluster/examples/kubernetes/ceph/create-external-cluster-resources.py:603]: wrong indent?

File: cluster/examples/kubernetes/ceph/create-external-cluster-resources.py [hunk 2]
```diff
@@ -357,18 +370,28 @@ def get_active_ceph_mgr(self):
...
+            # collect all the 'stand-by' mgr ips
+            mgr_ips = []
+            for each_standby_mgr in standby_mgrs:
+                failed_ip = each_standby_mgr
+                mgr_ips.append(
+                        self._convert_hostname_to_ip(each_standby_mgr))
         except:
             raise ExecutionFailureException(
-                "unable to convert a hostname to an IP address, monitoring host name: {}".format(monitoring_endpoint_ip))
+                "unable to convert a hostname to an IP address, for host: {}".format(failed_ip))
```
Comment (leseb) [inline cluster/examples/kubernetes/ceph/create-external-cluster-resources.py:384]: "please enter the IP addresses of all the ceph-mgr with the '--monitoring-endpoint' flag"

**Review by leseb — CHANGES_REQUESTED** [review]

File: cluster/examples/kubernetes/ceph/create-external-cluster-resources.py (hunk 1 above)
Comment (aruniiird) [reply cluster/examples/kubernetes/ceph/create-external-cluster-resources.py:603]: Not really. If you see the above line, we have a back-slash (\) , which means a continuation of the line. So we are continuing from 602 line.

**Comment by leseb** [comment]
@aruniiird see CI failures https://github.com/rook/rook/pull/7217/checks?check_run_id=1910532384

File: cluster/examples/kubernetes/ceph/create-external-cluster-resources.py (hunk 2 above)
Comment (aruniiird) [reply cluster/examples/kubernetes/ceph/create-external-cluster-resources.py]: done...

**Review by leseb — CHANGES_REQUESTED** [review]
You need remove this line

https://github.com/rook/rook/blob/371128224b28573fff06e7d850fd2cbb023df0a8/.github/workflows/canary-integration-test.yml#L86

And change that line https://github.com/rook/rook/blob/371128224b28573fff06e7d850fd2cbb023df0a8/.github/workflows/canary-integration-test.yml#L87 to point to `/etc/rook-external` instead.

File: cluster/examples/kubernetes/ceph/create-external-cluster-resources.py [hunk 3]
```diff
@@ -328,21 +334,27 @@ def _convert_hostname_to_ip(self, host_name):
...
         return ip
 
-    def get_active_ceph_mgr(self):
+    def get_active_and_standby_mgrs(self):
         monitoring_endpoint_port = self._arg_parser.monitoring_endpoint_port
         monitoring_endpoint_ip = self._arg_parser.monitoring_endpoint
+        standby_mgrs = []
         if not monitoring_endpoint_ip:
-            cmd_json = {"prefix": "mgr services", "format": "json"}
+            cmd_json = {"prefix": "status", "format": "json"}
```
Comment (leseb) [inline cluster/examples/kubernetes/ceph/create-external-cluster-resources.py:342]: With this, I think we need to update the --upgrade flag too right?

**Review by leseb — CHANGES_REQUESTED** [review]

File: cluster/examples/kubernetes/ceph/create-external-cluster-resources.py (hunk 3 above)
Comment (aruniiird) [reply cluster/examples/kubernetes/ceph/create-external-cluster-resources.py:342]: > With this, I think we need to update the --upgrade flag too right?

IMHO, we don't have to.
We have 3 sets of flags, `common`, 'output` and `upgrade` flags. The current changes we made (to add the standby mgrs also into the list of mgrs) is part of `output` flag, which is independent of `upgrade` flag (and don't affect the `upgrade` flag).
`upgrade` is mostly about upgrading the permissions/caps of the existing user.

Comment (leseb) [reply cluster/examples/kubernetes/ceph/create-external-cluster-resources.py:342]: Ah right, the caps are only needed for the operator, not the script, my bad!

File: cluster/examples/kubernetes/ceph/create-external-cluster-resources.py [hunk 4]
```diff
@@ -357,16 +369,28 @@ def get_active_ceph_mgr(self):
...
+            # collect all the 'stand-by' mgr ips
+            mgr_ips = []
+            for each_standby_mgr in standby_mgrs:
+                failed_ip = each_standby_mgr
+                mgr_ips.append(
+                        self._convert_hostname_to_ip(each_standby_mgr))
         except:
             raise ExecutionFailureException(
-                "unable to convert a hostname to an IP address, monitoring host name: {}".format(monitoring_endpoint_ip))
+                "Conversion of host: {} to IP failed. "\
```
Comment (leseb) [inline cluster/examples/kubernetes/ceph/create-external-cluster-resources.py:383]: Are you returning an error here? The CI fails but it looks like the script outputs 0. See: 

```
2s
Run kubectl -n rook-ceph exec $(kubectl get pod -l app=rook-ceph-tools -n rook-ceph -o jsonpath='{.items[0].metadata.name}') -- mkdir -p /etc/ceph/test-data
  kubectl -n rook-ceph exec $(kubectl get pod -l app=rook-ceph-tools -n rook-ceph -o jsonpath='{.items[0].metadata.name}') -- mkdir -p /etc/ceph/test-data
  kubectl -n rook-ceph cp cluster/examples/kubernetes/ceph/test-data/ceph-status-out $(kubectl get pod -l app=rook-ceph-tools -n rook-ceph -o jsonpath='{.items[0].metadata.name}'):/etc/ceph/test-data/
  kubectl -n rook-ceph cp cluster/examples/kubernetes/ceph/create-external-cluster-resources.py $(kubectl get pod -l app=rook-ceph-tools -n rook-ceph -o jsonpath='{.items[0].metadata.name}'):/etc/ceph
  kubectl -n rook-ceph exec $(kubectl get pod -l app=rook-ceph-tools -n rook-ceph -o jsonpath='{.items[0].metadata.name}') -- python3 /etc/ceph/create-external-cluster-resources.py --rbd-data-pool-name replicapool
  shell: /bin/bash -e {0}
  env:
    GOROOT: /opt/hostedtoolcache/go/1.15.8/x64
    MINIKUBE_HOME: /home/runner/work/_temp
Execution Failed: Conversion of host: rook-ceph-mgr-a-664bb58559-vdwqh to IP failed. Please enter the IP addresses of all the ceph-mgrs with the '--monitoring-endpoint' flag
```

Comment (aruniiird) [reply cluster/examples/kubernetes/ceph/create-external-cluster-resources.py:383]: I am raising a python exception here, but at the call-site, we are catching the exception thus returning with a ZERO exit value.
Fixed the issue in the latest CI fix commit.

Comment (leseb) [reply cluster/examples/kubernetes/ceph/create-external-cluster-resources.py:383]: What's the fix? Where is it?

**Comment by BlaineEXE** [comment]
Will this affect or be affected by https://github.com/rook/rook/pull/7235?

**Comment by leseb** [comment]
> Will this affect or be affected by #7235?

No, it's fine, #7235 is actually helping the external case too. I'll work on the external approach.

**Comment by leseb** [comment]
@aruniiird any updates?`

**Comment by aruniiird** [comment]
> @aruniiird any updates?`

Making changes @leseb ... will update today...

**Comment by aruniiird** [comment]
Thanks @subhamkrai to provide me a way to write into a file inside the pod. (=
@leseb , now the tests are passing (as the `mgr` hostname is added into the `/etc/hosts` file of `tools` pod)
Please take a look...

File: cluster/examples/kubernetes/ceph/create-external-cluster-resources.py (hunk 4 above)
Comment (aruniiird) [reply cluster/examples/kubernetes/ceph/create-external-cluster-resources.py:383]: At the above mentioned line, we are raising the error but at the calling place we are catching/handling the error. So it is terminating successfully. The fix was to re-raise the exception at the call site.
https://github.com/rook/rook/blob/bc4dd243069f5d3af7b04fb07ff489ece30ac78c/cluster/examples/kubernetes/ceph/create-external-cluster-resources.py#L800

**Review by leseb — APPROVED** [review]

**Comment by leseb** [comment]
@aruniiird  please submit a backport PR against the stable-1.5 branch. Thanks

**Pull request merged**